Enable GUI Visualization on Docker with X11 Forwarding

Jpfonseca opened the pull request:
This PR solves the issues identified in  #882 and #4885.
In specific @gadhane and me have investigated the issues related with visualizing the GUI when running the docker image with X11 Forwarding in Ubuntu.

The issue and details are thoroughly discussed  [here](https://github.com/ultralytics/ultralytics/issues/882#issuecomment-2096151026).


<!--
Thank you 🙏 for your contribution to [Ultralytics](https://ultralytics.com) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. **Check for Existing Contributions**: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. **Link Related Issues**: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. **Elaborate Your Changes**: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. **Ultralytics Contributor License Agreement (CLA)**: To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    _I have read the CLA Document and I sign the CLA_

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing). Your adherence to these guidelines ensures a faster and more effective review process.
--->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements in Docker compatibility and improved display checks in environment.

### 📊 Key Changes
- Added `libsm6` installation to the Dockerfile, essential for creating QT-based windows for object visualization.
- Modified environment check logic in `checks.py` for better compatibility and error handling related to displaying images.

### 🎯 Purpose & Impact
- 🚀 **Enhanced Visualization Support**: Installing `libsm6` allows Docker environments to more easily visualize detection results, aiding in debugging and development processes.
- 🛠 **Improved User Experience**: Modified display checks now more accurately identify when and why image displays might fail, leading to clearer error messages and less confusion for users setting up or debugging their environments.
- ✅ **Broader Compatibility**: These changes improve the software’s compatibility with various platforms, making it more accessible and easier to use across different setups.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements to Docker setup and GUI support for Ultralytics.

### 📊 Key Changes
- **Dockerfile Update**: Added `libsm6` for creating QT-based windows, enabling visualization in Docker containers. 🐳
- **Docker Quickstart Guide Enhancements**: Expanded with instructions for running Ultralytics using CPU or GPU, and incorporating GUI applications in Docker containers. 📖
- **Experimental GUI Support**: Provided instructions for displaying Ultralytics detection results through GUI within a Docker container, marking a significant step towards more interactive usage scenarios. ⚠️🖥️
- **Check Update in Utils**: Simplified conditions for the environment's capability to display images, specifically refining the checks for `DISPLAY` environment variable presence. 🛠️

### 🎯 Purpose & Impact
- **Enhanced Visualization**: The inclusion of `libsm6` and detailed guide for GUI support allows users to visualize object detection results directly, significantly improving the user experience for those working in Docker environments. 🎉
- **Greater Flexibility**: By addressing CPU and GPU usage more explicitly, and offering guidance on utilizing GUI, users gain more flexibility in how they deploy and interact with Ultralytics software, catering to a wider range of use cases. 🌈
- **Improved Accessibility**: Simplifying environment check conditions makes it easier for developers to troubleshoot and ensure their setup is capable of visual output, streamlining the setup process. 🚀

This PR facilitates a more interactive and user-friendly approach to using Ultralytics in varied environments, making it accessible to both developers and users interested in advanced object detection without sacrificing the convenience of Docker.

Jpfonseca committed:
Adds dependency to ensure that cv2.imshow() functions correctly when using QT

This change adds the libsm6 package as a dependency of the docker container builds. This ensures cv2.imshow() functions correctly when using QT, as  this package is one of the requirements of libqxcb.

Jpfonseca committed:
Updates checks.py to allow GUI functions like cv2.imshow() when $DISPLAY is properly et in docker environment

This change removes the docker-specific assertion, enabling the GUI functions to run in the docker container. To enable this support, the $DISPLAY environment variable must be properly set, as well as the X11Forwarding needs to be properly configured.

Jpfonseca commented:
I have read the CLA Document and I sign the CLA

ambitious-octopus reviewed (commented):
Seems like a good solution. However, I would also add to the documentation the specific arguments for launching the container in order to connect to the X server.

glenn-jocher committed:
Simplify update

glenn-jocher commented:
@ambitious-octopus hey that's a great idea!

@Jpfonseca can you add an example to the Docker Quickstart Guide here with the code to visualize from a Docker image?
https://docs.ultralytics.com/guides/docker-quickstart/
https://github.com/ultralytics/ultralytics/blob/main/docs/en/guides/docker-quickstart.md

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher commented:
@Jpfonseca PR looks but could you please update the docs with example code for remote visualization here:
https://github.com/ultralytics/ultralytics/blob/main/docs/en/guides/docker-quickstart.md

glenn-jocher committed:
Merge branch 'main' into main

Jpfonseca committed:
Added examples for remote visualization in linux environments

Added information on to properly visualize the Object detection from a docker container in both GNU-Linux Display Servers (Xorg and Wayland )

Co-Authored-By: Gereziher Adhane <gereziherw@gmail.com>

Jpfonseca commented:
@glenn-jocher just added the examples for both X11 and Wayland. Tested both solutions in Xorg (gnome and xfce) and Wayland (sway and gnome) .

ambitious-octopus commented on docs/en/guides/docker-quickstart.md:
````diff
@@ -103,6 +108,52 @@ sudo docker run -it --ipc=host --gpus '"device=2,3"' $t
+#### Wayland
+1. Open a terminal and use the following command to run the container:
+```bash
+docker run --env DISPLAY=$DISPLAY --volume $XDG_RUNTIME_DIR/$WAYLAND_DISPLAY:/tmp/$WAYLAND_DISPLAY -e QT_DEBUG_PLUGINS=1 --net=host -it $t  
````
```suggestion
docker run -e DISPLAY=$DISPLAY -v $XDG_RUNTIME_DIR/$WAYLAND_DISPLAY:/tmp/$WAYLAND_DISPLAY -e QT_DEBUG_PLUGINS=1 --net=host -it $t  
```
Change the format of the options to be consistent with the previous command (X11).

ambitious-octopus replied:
Same here with `QT_DEBUG_PLUGINS=1`.

ambitious-octopus commented on docs/en/guides/docker-quickstart.md:
````diff
@@ -103,6 +108,52 @@ sudo docker run -it --ipc=host --gpus '"device=2,3"' $t
+1. Open a terminal and run:
+
+```bash
+docker run -e DISPLAY=$DISPLAY -e QT_DEBUG_PLUGINS=1  -v /tmp/.X11-unix:/tmp/.X11-unix -v ~/.Xauthority:/root/.Xauthority -it $t
````
Is it necessary to set the variable `QT_DEBUG_PLUGINS=1` here? We risk filling the terminals with debug messages.

ambitious-octopus commented on docs/en/guides/docker-quickstart.md:
```diff
@@ -103,6 +108,52 @@ sudo docker run -it --ipc=host --gpus '"device=2,3"' $t
 
+### Visualize Ultralytics' Object detection in your GNU-Linux Display Server
+
+The following instructions are highly experimental. Sharing a X11 socket with a docker container can raise security concerns, and as such this solution should be tested only in a controlled environment. 
```
Here I would add a small box, with a link with more information on security risks, perhaps a link to a resource. 

Furthermore I would also add that when you are done using the container you can use `xhost -local:root` to restore access control.

ambitious-octopus replied:
You can create a warning box like this:

```
???+ warning "Something"
    Text Here
```

![image](https://github.com/ultralytics/ultralytics/assets/3855193/2812b457-3a08-4a3a-b3b5-f9657a19359f)

Jpfonseca replied:
Thanks. I am new to that kind of formatting ;D

ambitious-octopus commented on docs/en/guides/docker-quickstart.md:
```diff
@@ -103,6 +108,52 @@ sudo docker run -it --ipc=host --gpus '"device=2,3"' $t
+
+A simple way to validate that the docker group has access to the X11 server is to run a container with a GUI program like `xclock` or `xeyes`. Alternatively, you can also install these programs in the ultralytics docker container to test the access to the X11 Server of you GNU-Linux Display Server .
+
+Below there are presented two examples on how to run the Ultralytics Object detection container in  in both Xorg and Wayland in interact mode `-it`. You can run the commands to start the docker container with using a [GPU](#using-gpus)
```
some typos here! :smile:

ambitious-octopus commented on docs/en/guides/docker-quickstart.md:
````diff
@@ -103,6 +108,52 @@ sudo docker run -it --ipc=host --gpus '"device=2,3"' $t
+```
+This allows connections from the `docker` group into the X11 Server. 
+
+A simple way to validate that the docker group has access to the X11 server is to run a container with a GUI program like `xclock` or `xeyes`. Alternatively, you can also install these programs in the ultralytics docker container to test the access to the X11 Server of you GNU-Linux Display Server .
````
This part can be replaced with: `yolo predict show=True`

Jpfonseca replied:
I believe it is better to have both. As 2 issues might arise. You might face issues with the X11Server and that can be tested with both. However if you also have issues with QT you are not able to render the object detection.

ambitious-octopus reviewed (changes requested):
Thank you for your contribution. I appreciate your input, as it adds significant value to the docs. I'll integrate this section more seamlessly, ensuring clarity and accuracy. Additionally, I'll correct any typos. Thanks for including instructions on using Docker without a GPU. It would also be useful to make a small integration here:

https://github.com/ultralytics/ultralytics/blob/fcfc44ea9cf529685e0afaba6e62f7d2d40c84d8/docs/en/guides/docker-quickstart.md?plain=1#L17-L32

Jpfonseca committed:
Update docs/en/guides/docker-quickstart.md

Co-authored-by: Francesco Mattioli <Francesco.mttl@gmail.com>

Jpfonseca committed:
Addressed documentation comments

Co-Authored-By: Gereziher Adhane <gereziherw@gmail.com>

Jpfonseca commented:
@ambitious-octopus we addressed the comments you made.

ambitious-octopus commented on docs/en/guides/docker-quickstart.md:
````diff
@@ -103,6 +109,62 @@ sudo docker run -it --ipc=host --gpus '"device=2,3"' $t
+xhost -local:docker
+```
+
+A simple way to validate that the docker group has access to the X11 server is to run a container with a GUI program like `xclock` or `xeyes`. Alternatively, you can also install these programs in the ultralytics docker container to test the access to the X11 Server of your GNU-Linux Display Server .
+
+Below, two examples are presented on how to run the Ultralytics Object detection container in both Xorg and Wayland using the interact mode `-it`. You can run the commands to start the docker container using a [GPU](#using-gpus)
+#### Xorg
+1. Open a terminal and run:
+
+```bash
+#Note 1: To debug the QT Plugins you can add the following flag to the command: -e QT_DEBUG_PLUGINS=1 
+#Note 2: Wayland also supports running the Xorg command.
+docker run -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix -v ~/.Xauthority:/root/.Xauthority -it $t
+```
+2. Open another terminal and run:
+```bash
+xhost +local:docker
+```
+3. Test the connection to the X11 Server inside the docker terminal:
+```bash
+yolo predict show=True
+```
+#### Wayland
+1. Open a terminal and use the following command to run the container:
+```bash
+# To debug the QT Plugins you can add the following flag to the command: -e QT_DEBUG_PLUGINS=1 
+docker run -e DISPLAY=$DISPLAY -v $XDG_RUNTIME_DIR/$WAYLAND_DISPLAY:/tmp/$WAYLAND_DISPLAY --net=host -it $t  
+
+```
+
+2. Open another terminal and run:
+```bash
+xhost +local:docker
+#Note: You may need to rerun the command each time that the X11 Server session closes. 
+```
+3. Test the connection to the X11 Server inside the docker terminal:
+```bash
+yolo predict show=True
+```
+
````
```suggestion

        ## Running Ultralytics in Docker Container
        
        Here's how to execute the Ultralytics Docker container:
        ### Using only the CPU
        ```bash
        # Run with all GPUs
        sudo docker run -it --ipc=host $t
        ```
        ### Using GPUs
        ```bash
        # Run with all GPUs
        sudo docker run -it --ipc=host --gpus all $t
        
        # Run specifying which GPUs to use
        sudo docker run -it --ipc=host --gpus '"device=2,3"' $t
        ```
        
        The `-it` flag assigns a pseudo-TTY and keeps stdin open, allowing you to interact with the container. The `--ipc=host` flag enables sharing of host's IPC namespace, essential for sharing memory between processes. The `--gpus` flag allows the container to access the host's GPUs.
        
        ### Run graphical user interface (GUI) applications in a Docker Container
        
        ???+ danger "Caution: Highly Experimental"
        
            The following instructions are experimental. Sharing a X11 socket with a docker container poses potential security risks. Therefore, it's recommended to test this solution only in a controlled environment. For more information, refer to this [resource](http://users.stat.umn.edu/~geyer/secure.html).
        
        Docker is primarily used to containerize background applications and CLI programs, but it can also run graphical programs. In the Linux world, two main graphic servers handle graphical display: X11 and Wayland. Before starting it's essential to determine which graphics server you're currently using. Just run this command to find out:
        
        ```bash
        env | grep -E -i 'x11|xorg|wayland'
        ```
        
        !!! example ""
        
            === "X11"
        
                If you're using X11, you can run the following command to allow the Docker container to access the X11 socket:
        
                ```bash
                xhost +local:docker && docker run -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix -v ~/.Xauthority:/root/.Xauthority -it --ipc=host $t
        
                ```
        
                This command sets the `DISPLAY` environment variable to the host's display, mounts the X11 socket, and maps the `.Xauthority` file to the container. The `xhost +local:docker` command allows the Docker container to access the X11 server.
        
                ??? info "Use GPUs"
                    If you're using [GPUs](#using-gpus), you can add the `--gpus all` flag to the command.
        
        
            === "Wayland"
        
                For Wayland, use the following command:
        
                ```bash
                xhost +local:docker && docker run -e DISPLAY=$DISPLAY -v $XDG_RUNTIME_DIR/$WAYLAND_DISPLAY:/tmp/$WAYLAND_DISPLAY --net=host -it --ipc=host $t
                ```
        
                This command sets the `DISPLAY` environment variable to the host's display, mounts the Wayland socket, and allows the Docker container to access the Wayland server.
        
                ??? info "Use GPUs"
                    If you're using [GPUs](#using-gpus), you can add the `--gpus all` flag to the command.
        
        Now you can display graphical applications inside your docker container. For example, you can run the following command to visualize the predictions of the YOLO model:
        
        ```bash
        yolo predict show=True
        ```
        
        ???+ warning "Revoke access"
        
            In both cases, don't forget to revoke access from the docker group when you're done.
            ```bash
            xhost -local:docker
            ```
        
        ??? info "Testing"
        
            A simple way to validate that the docker group has access to the X11 server is to run a container with a GUI program like `xclock` or `xeyes`. Alternatively, you can also install these programs in the ultralytics docker container to test the access to the X11 server of your GNU-Linux Display Server. If you run into any problems, consider setting the environment variable `-e QT_DEBUG_PLUGINS=1`. This action enables the output of debugging information, aiding in the troubleshooting process.
```

Jpfonseca replied:
Yes I have validated the commands , in my personal setup sway (wayland) setup and another machine running ubuntu (Xorg).

Jpfonseca replied:
@ambitious-octopus  these are the specs of the machine: 
<details>
![image](https://github.com/ultralytics/ultralytics/assets/11836470/c7c5048a-286d-44f2-99f7-8cb801e525d8)
</details>

Jpfonseca replied:
Your suggestion looks good. I will wait for  @Burhan-Q input before I commit the suggestion.

ambitious-octopus reviewed (changes requested):
What do you think of this solution? I tried to follow what you wrote but made it more concise and used the ultralytics documentation standard. You can test it with `mkdocs serve`. What do you think @Burhan-Q? 

Also, I haven't tested with wayland. Have you already tested @Jpfonseca? With what machine?

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher committed:
Merge branch 'main' into main

Burhan-Q commented:
@Jpfonseca and @ambitious-octopus a few notes in general on these additions/changes. Expand the section at the bottom of this comment to see a preview of the changes.

- Let's make the Visualization it's own section using an `H2` header

- Double check alignment with proposed changes from @ambitious-octopus as they are in the GitHub preview, the alignment is off.

- Consider the section title `Using a Display Server with Docker to Show Ultralytics Detection Results` instead of `Visualize Ultralytics' Object detection in your GNU-Linux Display Server` as Ultralytics can be used for more than object detection and is more generally descriptive (although still more verbose than I like).

    - Also generally, use `Ultralytics` instead of `Ultralytics'` (no trailing apostrophe) as it's not required

- Make certain to always capitalize Ultralytics and Docker whenever used in text. If referring to the Ultralytics package, you can use `ultralytics`

- Make sure to always include new lines before and after code blocks.

        In both cases, don't forget to revoke access from the docker group when you're done.

        ```bash
        xhost -local:docker
        ```
    
        Next line starts here...

-  `???+ danger "Caution: Highly Experimental"` is a good change, however I'd recommend `!!! danger "Highly Experimental - User Assumes All Risk"` but at the very least, don't think it should be collapsible (change from `???+` to `!!!`); expand section at end of comment to see preview.

- It would be good to see more specifics about what the user should be looking to learn/understand from `More information on the subject [here](http://users.stat.umn.edu/~geyer/secure.html)` as it's not clear what a user should look to understand specifically (all commands from the link, 1 command, or something else?).

    - I also think it might be good to use a source that we can rely on. If looking to point users to these specific commands, perhaps using [ssh](https://linux.die.net/man/1/ssh), [scp](https://linux.die.net/man/1/scp), [xauth](https://linux.die.net/man/1/xauth), and [xhost](https://linux.die.net/man/1/xhost)

- Be sure to include an example of "what you should see and what it means" after instructing to use
  
  ```bash
  env | grep -E -i 'x11|xorg|wayland'
  ```

since it's not obvious what a user should expect to see based on what's available (or when nothing is available).

- I think the tabbed example can be changed (expand section at end of comment to see preview)

```diff
- !!! example ""
+ !!! example
```

- I like the tabbed examples, but maybe break up the lines with `\` so there's no need to scroll to view the entire command (expand section at end of comment to see preview)

    ```bash
    xhost +local:docker && docker run -e DISPLAY=$DISPLAY -v \
    $XDG_RUNTIME_DIR/$WAYLAND_DISPLAY:/tmp/$WAYLAND_DISPLAY \
    --net=host -it --ipc=host $t
    ```
    
<details><summary>Preview single line with horizontal scrolling</summary>
<p>

![image](https://github.com/ultralytics/ultralytics/assets/62214284/376613a8-9336-48c9-ac19-5c40d3f242c4)

</p>
</details> 

- I think this block can be moved to it's own section (above) the tabbed examples so that it's visible for both tabs and doesn't need to be duplicated  (expand section at end of comment to see preview):

    ```md
    !!! example ""

        ??? info "Use GPUs"
        
            If you're using [GPUs](#using-gpus), you can add the `--gpus all` flag to the command.
        
            === "X11"
        
                If you're using X11, you can run the following command to allow the Docker container to access the X11 socket:
            
                ...
    ```

- Make this non-collapsible as I think it's important for users to always see (expand section at end of comment to see preview)

    ```diff
    - ???+ warning "Revoke access"
    + !!! warning "Revoke access"
    ```

- I'll admit I'm biased here (I wrote the guide), but I think it would be useful to mention the `view-results-in-terminal.md` guide as it's somewhat related. I think this could be good to include in the introduction to this section

<details><summary>Preview of suggestions</summary>
<p>

## Table of Contents

![image](https://github.com/ultralytics/ultralytics/assets/62214284/2f5afa11-6107-4d85-8670-3454de4bd555)

## Section

![image](https://github.com/ultralytics/ultralytics/assets/62214284/7afb9faa-999d-4ea6-843b-7c5046a3ca4c)

</p>
</details>

Burhan-Q committed:
Merge branch 'main' into main

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher committed:
Merge branch 'main' into main

Jpfonseca committed:
Apply suggestions from code review

Co-authored-by: Francesco Mattioli <Francesco.mttl@gmail.com>

Jpfonseca committed:
Update the guide

Updated the guide following the suggestions by @Burhan-Q and  @ambitious-octopus.

Co-Authored-By: Burhan <62214284+Burhan-Q@users.noreply.github.com>
Co-Authored-By: Gereziher Adhane <gereziherw@gmail.com>
Co-Authored-By: Francesco Mattioli <Francesco.mttl@gmail.com>

Jpfonseca committed:
Merge branch 'main' into main

glenn-jocher commented:
@Burhan-Q is this PR good to merge now?

glenn-jocher committed:
Merge branch 'main' into main

Burhan-Q committed:
capitalize all text references to Docker

Burhan-Q committed:
reformat citations and links

Burhan-Q committed:
fix warning "Revoke access" (not rendering), now not collapsible and new line before code bloc

Burhan-Q committed:
line clean ups

Burhan-Q committed:
Reorder sections, add sub-headers, move reference to view images in terminal guide, and add more docs reference links

Burhan-Q committed:
add external reference links and statement about display server setup

Burhan-Q reviewed (approved):
@glenn-jocher I went ahead and made a few changes, fixes. I believe it's good to go.

- capitalize all text references to Docker
- reformat citations and links
- fix warning "Revoke access" (not rendering), now not collapsible, and new line before code block
- line spacing clean up
- Reorder sections, add sub-headers, move reference to view images in terminal guide, and add more docs reference links
- add external reference links and statement about display server setup

<details><summary>Page Preview</summary>
<p>

![image](https://github.com/ultralytics/ultralytics/assets/62214284/f7a71fc0-ac41-46d3-a0f6-77da67ecc822)

</p>
</details>

Burhan-Q commented:
Also, I can't speak to the functionality of this guide and have assumed it to be correct.

ambitious-octopus reviewed (commented):
Tested with `x11`, everything works fine. I'm testing with `wayland`, I'll confirm asap!

Jpfonseca commented:
@ambitious-octopus  Thanks, if needed, I can record it now .

ambitious-octopus reviewed (commented):
It might be useful, thanks @Jpfonseca !

Jpfonseca commented:
Here is the video as requested by @ambitious-octopus. 
 I started the session with `xhost +local:docker ` enabled, but afterwards I revert the command. and use the command again to validate the access to  the `$DISPLAY`

https://github.com/ultralytics/ultralytics/assets/11836470/c1a2d0c2-395f-4f17-a9e2-16a9225e13b9

ambitious-octopus reviewed (commented):
Hey @Jpfonseca thanks for the video, everything seems to work correctly!

Jpfonseca commented:
@glenn-jocher  I believe it is ready to be merged now.

glenn-jocher committed:
Merge branch 'main' into main

glenn-jocher commented:
@Jpfonseca @ambitious-octopus @Burhan-Q awesome, thank you all, PR merged!

Jpfonseca commented:
@glenn-jocher  @ambitious-octopus  @Burhan-Q if needed, i can also share how this can  be used in a K8s  Deployment in a different PR

Burhan-Q commented:
@Jpfonseca you never have to ask permission to open a PR 👍 however I will only be able to provide support from the docs formatting perspective, as I have only gotten as far as learning `k8s == kubernetes` 😆
